### PR TITLE
Validation added for Target resource type for include and revinclude search options

### PR DESCRIPTION
**Pull request by rajithaalurims**
## Description
Invalid Target resource type was not handled when search with "include" and "revinclude" options. 

## Related issues
[76141](https://microsofthealth.visualstudio.com/Health/_workitems/edit/76141)

## Testing
**Test 1:** 
Test URL : Patient/?revinclude=Observation:subject:NotAResourceType
Returned below error:
"issue": [
        {
            "severity": "error",
            "code": "not-supported",
            "diagnostics": "Resource type 'NotAResourceType' is not supported.'."
        }
    ]

**Test 2:** 
Test URL : Patient/?include=Observation:subject:NotAResourceType
Returned below error:
"issue": [
        {
            "severity": "error",
            "code": "not-supported",
            "diagnostics": "Resource type 'NotAResourceType' is not supported.'."
        }
    ]

## FHIR Team Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [ ] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [ ] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [ ] CI is green before merge
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs [hunk 1]
```diff
@@ -416,6 +416,12 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
                             string.Format(Core.Resources.RevIncludeIterateTargetTypeNotSpecified, p.query));
                     }
 
+                    if (expression.TargetResourceType != null &&
```
Comment (rbans96) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:419]: what if its empty or whitespace?

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/IncludeSearchTests.cs [hunk 2]
```diff
@@ -1352,6 +1352,21 @@ public async Task GivenARevIncludeIterateSearchExpressionWithCircularReference_W
             };
         }
 
+        [Fact]
+        [HttpIntegrationFixtureArgumentSets(DataStore.SqlServer)]
+        public async Task GivenARevIncludeIterateSearchExpressionWithInvalidTargetResourceType_WhenSearched_ShouldThrowResourceNotSupportedException()
+        {
+            string query = $"_include=Observation:subject:NotAResourceType";
```
Comment (rbans96) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/IncludeSearchTests.cs:1359]: Should we add a test for _include=Observation:subject:

Comment (rajithaalurims) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/IncludeSearchTests.cs:1374]: Added test cases for empty/white space

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs [hunk 3]
```diff
@@ -421,6 +422,17 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
...
+                       string.IsNullOrWhiteSpace(expression.TargetResourceType))
+                    {
+                        throw Error.Format($"Invalid TargetResourceType : value cannot be empty");
+                    }
+
+                    if (expression.TargetResourceType != null && !ModelInfoProvider.IsKnownResource(expression.TargetResourceType))
```
Comment (rbans96) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:431]: nit: Should we remove this check from here as its already handled above
expression.TargetResourceType != null

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/IncludeSearchTests.cs [hunk 4]
```diff
@@ -1352,6 +1352,96 @@ public async Task GivenARevIncludeIterateSearchExpressionWithCircularReference_W
...
+
+            string[] expectedDiagnostics = { string.Format("Invalid TargetResourceType : value cannot be empty") };
+            IssueSeverity[] expectedIssueSeverities = { IssueSeverity.Error };
+            IssueType[] expectedCodeTypes = { IssueType.Invalid };
+            ValidateOperationOutcome(expectedDiagnostics, expectedIssueSeverities, expectedCodeTypes, fhirException.OperationOutcome);
+        }
+
+        [Fact]
+        [HttpIntegrationFixtureArgumentSets(DataStore.SqlServer)]
+        public async Task GivenAIncludeIterateSearchExpressionWithWhiteSpaceTargetResourceType_WhenSearched_ShouldThrowErrorFormat()
```
Comment (rbans96) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/IncludeSearchTests.cs:1402]: nit: Should we use InlineData for empty and whitespace tests to combine them in one?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs (hunk 3 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:431]: First condition is to handle empty or white spaces.
Second condition is to validate if given Target resource type is invalid.

File: src/Microsoft.Health.Fhir.Core/Resources.resx [hunk 5]
```diff
@@ -243,6 +243,10 @@
     <value>The '_revinclude:iterate={0}' search parameter has multiple target types. Please specify a target type.</value>
     <comment>_revinclude:iterate is a query parameter.</comment>
   </data>
+  <data name="IncludeRevIncludeInvalidTargetResourceType" xml:space="preserve">
+    <value>Invalid TargetResourceType : value cannot be empty.</value>
```
Comment (rbans96) [inline src/Microsoft.Health.Fhir.Core/Resources.resx:247]: nit: Can we simplify the error message - 'The target resource type cannot be empty'

**Review by rbans96 — DISMISSED** [review]

**Review by rbans96 — APPROVED** [review]

File: src/Microsoft.Health.Fhir.Core/Resources.resx (hunk 5 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.Core/Resources.resx]: yes

**Pull request merged**